### ISSUE TITLE
[bug] Meson does not respect buildenv PATH

### ISSUE THREAD

**Issue by duncanspumpkin** [conan-io/conan#13936]
### Environment details

* Operating System+version: Windows 10
* Compiler+version: all
* Conan version: 1.57
* Python version: all

When using meson toolchain anything added to the buildenv path does not appear to be respected. I have python installed to a custom location so i need to ensure that meson is run with the correct path variables otherwise it will fail to find python. To achieve that I would normally just add pythons path to [buildenv] PATH.

I made a temporary work around locally where https://github.com/conan-io/conan/blob/2a17403626451b7f62cd77eccba80664e1f3d0c8/conan/tools/meson/meson.py#L29 is modified to call "meson.py" (instead of "meson.cmd" which is a wrapper that calls python3 meson.py). This does sort of work but is messy it would be better if the build env was respected.

### Steps to reproduce

1. Try build a meson package on Windows with python not in your PATH.
2. Will fail.

**Comment by memsharded** [comment]
Hi @duncanspumpkin 

I am having a look at this.
I am trying with this recipe:
```python
from conan import ConanFile

class Pkg(ConanFile):
    name = "pkg"
    version = "0.1"
    generators = "VirtualBuildEnv"

    def build(self):
        self.run("python --version")
        self.run("meson --version")
```
and this profile:
```
[buildenv]
PATH=+(path)c:/ws/Python/Python311
```
Then
```
$ conan create . -pr=profile
...
Python 3.11.3
0.61.2
```
seems to be executing the ``run()`` commands with Python3.11 (my default while executing is other not 3.11)

**Comment by duncanspumpkin** [comment]
The issue is with the meson toolchain I think the following should be short enough to trigger it. Unless are you saying I shouldn't be using the meson toolchain and should call meson directly?
 
```python
from conans import ConanFile, tools
from conan.tools.gnu import PkgConfigDeps
from conan.tools.env import VirtualBuildEnv, Environment
from conan.tools.meson import Meson, MesonToolchain
from conan.tools.files import apply_conandata_patches, copy, export_conandata_patches, get
import os

class ConanFileToolsTest(ConanFile):
    name = "test"
    settings = "os", "arch", "compiler", "build_type"

    def build_requirements(self):
        self.tool_requires("meson/1.0.0")

    def layout(self):
        self.folders.build = "build"

    def generate(self):
        deps = PkgConfigDeps(self)
        deps.generate()
        tc = MesonToolchain(self)
        tc.generate()

    def build(self):
        apply_conandata_patches(self)
        meson = Meson(self)
        meson.configure()
        meson.build()

    def package(self):
        meson = Meson(self)
        meson.install()
```

```
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
test/23.0.0:
test/23.0.0: ERROR: Package '48a8723658aa0b8a0afa6d77d3d7d3d8253860d3' build failed
test/23.0.0: WARN: Build folder C:\.conan\5f4864\1\build
ERROR: test/23.0.0: Error in build() method, line 65
        meson.configure()
```

I modified also meson to output what it has in path and its missing my additions to path. It does have all my tool_requires though added to the path.

**Comment by memsharded** [comment]
Seems to be working fine here, running Conan with Python3.6, with your ``conanfile.py`` and the profile:
```
include(default)

[buildenv]
PATH=+(path)c:/ws/Python/Python311
```


- When not using the profile.
```bash
conan build .
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
Meson works correctly only with python 3.7+.
You have python 3.6.8 (tags/v3.6.8:3c6b436a57, Dec 24 2018, 00:16:47) [MSC v.1916 64 bit (AMD64)].
Please update your environment
```
- When using the profile.
```bash
conan build . -pr=profile
conanfile.py (test/None): RUN: meson compile -C "C:\Users\memsharded\conanws\kk\build" -j8
conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - vcvars_ver=14.3
[vcvarsall.bat] Environment initialized for: 'x64'
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: C:\Users\memsharded\.conan2\p\ninjae2ad385cd85df\p\bin\ninja.EXE -j 8
[0/1] Regenerating build files.The Meson build system
Version: 1.0.0
Source dir: C:\Users\memsharded\conanws\kk
Build dir: C:\Users\memsharded\conanws\kk\build
Build type: native build
Project name: pkg
Project version: undefined
C++ compiler for the host machine: cl (msvc 19.35.32217.1 "Microsoft (R) C/C++ Optimizing Compiler Version 19.35.32217.1 for x64")
C++ linker for the host machine: link link 14.35.32217.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Build targets in project: 1

pkg  undefined
```

**Comment by duncanspumpkin** [comment]
Okay thanks for checking I tried again today and I can't reproduce my issue. What is odd though is I modified meson.cmd to add an `echo %PATH%` and in the resultant output I can't see any of the additions to path from the build env variable (I do see the additions to path for any tool_requires though). Maybe my understanding of how the environment is passed to sub processes is wrong.

Also I did notice i got a few issues if I tried to run things from a powershell developer prompt but thats unrelated to this.

**Comment by memsharded** [comment]
Then maybe we can close this issue? Is there something else we could help with? Thanks!

**Comment by duncanspumpkin** [comment]
I think i found the issue it looks like its tools.env.virtualenv:auto_use does not auto generate a build environment for meson toolchain. At least that appears to be happening. Perhaps this is fixed in v2 since auto_use is removed but in the v1.57 it doesn't appear to work.  Unfortunately I can't switch to conan v2 as its incompatible with our setup so I can't confirm my guess. After adding a manual `ms = VirtualBuildEnv(self)       ms.generate()` to the generate function meson is now showing the correct PATH and finding my build environment.

**Comment by memsharded** [comment]
Good to know that you found the issue and that explicit ``VirtualBuildEnv`` made it work.
I don't know why ``auto_use`` will not do it, its purpose is exactly that, the code is quite straightforward:
```python
    if conanfile.conf["tools.env.virtualenv:auto_use"]:
        mkdir(path)
        with chdir(path):
            if conanfile.virtualbuildenv:
                from conan.tools.env.virtualbuildenv import VirtualBuildEnv
                env = VirtualBuildEnv(conanfile)
                env.generate()
```
It is true that we have found some other tooling (Meson) and recipes to internally abuse ``VirtualBuildEnv``, so we had to protect it with:
```python
         prev_status = self._conanfile.virtualbuildenv
        build_env = VirtualBuildEnv(self._conanfile).vars()
        self._conanfile.virtualbuildenv = prev_status
```
It is very possible that this is included in one of the latest 1.X releases, so updating to 1.60 might fix it (I'd suggest to try it).

Regarding your Conan 2.0 readiness, while there will be time for the upgrade and 1.X will be supported, it is important to start doing the work to be ready, so please don't hesitate to reach out or create tickets with any question or doubt around it.

**Comment by duncanspumpkin** [comment]
Will try out the newer 1.60 and see if it fixes anything.

> Regarding your Conan 2.0 readiness, while there will be time for the upgrade and 1.X will be supported, it is important to start doing the work to be ready, so please don't hesitate to reach out or create tickets with any question or doubt around it.

Unfortunately its due to gitlab support not for any other reason (at least that's the major show stopper) https://gitlab.com/gitlab-org/gitlab/-/issues/389216
Might have to look into hosting a separate package repo but hopefully not.

**Comment by memsharded** [comment]
> Might have to look into hosting a separate package repo but hopefully not.

If you have to, recall that ArtifactoryCE is free, you can get it from https://conan.io/downloads.html, and it is not complicated to run it in your own server. I know it is not as convenient as having it integrated and already available, but if Gitlab doesn't prioritize support for v2 enough, at least you have this option.

**Comment by duncanspumpkin** [comment]
Ahhh: #13344 #13346 looks like you fixed this for 2.0.
